README.md
# Editor Enhancements
HTML and URL encodings, transformations and sorting of selected text the editor.

![Demo](https://raw.githubusercontent.com/gurayyarar/EditorEnhancements/master/images/demo.gif)

## Features
- HTML Encode/Decode
- URL Encode/Decode
- To Title Case
- To Lower Case
- To Upper Case
- Capitalize First Letter
- Reverse String
- Remove Accents/Diacritics
- MD5, SHA-1, SHA-256, SHA-384, SHA-512 Hashing
- Line Sorting (Ascending/Descending)
- Remove Empty Lines
- Remove Duplicate Lines
- Remove Horizontal White Space

## License
**Editor Enhancements** is an open source project that is licensed under the [MIT license](http://opensource.org/licenses/MIT).


## Donations
Donations are **greatly appreciated!**

**[BUY ME A COFFEE](http://bit.ly/2NUcFly)**

CHANGELOG.md
# Change Logs

- Initial Release **(July 29th, 2018)**
- Packages has been updated **(May 26, 2019)**

src/utils.ts
import * as crypto from "crypto";
const Entities = require('html-entities').AllHtmlEntities;
var removeDiacritics = require('diacritics').remove;
var titleCase = require('title-case');

const entities = new Entities();

export const Utils: any = {
    htmlEncode: (text: string): string => {
        return entities.encode(text);
    },
    htmlDecode: (text: string): string => {
        return entities.decode(text);
    },
    urlEncode: (text: string): string => {
        return encodeURIComponent(text);
    },
    urlDecode: (text: string): string => {
        return decodeURIComponent(text);
    },
    toTitleCase: (text: string): string => {
        return titleCase(text);
    },
    toLowerCase: (text: string): string => {
        return text.toLowerCase();
    },
    toUpperCase: (text: string): string => {
        return text.toUpperCase();
    },
    capitalizeFirstLetter: (text: string): string => {
        text = text.toLowerCase();
        return text.charAt(0).toUpperCase() + text.slice(1);
    },
    reverse: (text: string): string => {
        return text.split("").reverse().join("");
    },
    removeAccent: (text: string): string => {
        return removeDiacritics(text);
    },
    toMd5: (text: string): string => {
        return crypto.createHash("md5").update(text).digest("hex");
    },
    toSha1: (text: string): string => {
        return crypto.createHash("sha1").update(text).digest("hex");
    },
    toSha256: (text: string): string => {
        return crypto.createHash("sha256").update(text).digest("hex");
    },
    toSha384: (text: string): string => {
        return crypto.createHash("sha384").update(text).digest("hex");
    },
    toSha512: (text: string): string => {
        return crypto.createHash("sha512").update(text).digest("hex");
    },
    sortAscending: (text: string): string => {
        let pieces: string[] = text.split("\n");
        pieces = pieces.filter(function (n) { return n !== "" && n !== null });

        pieces = pieces.sort();
        return `${pieces.slice(0, -1).join('\n')}\n${pieces.slice(-1)}`;
    },
    sortDescending: (text: string): string => {
        let pieces: string[] = text.split("\n");
        pieces = pieces.filter(function (n) { return n !== "" && n !== null });

        pieces = pieces.sort().reverse();
        return `${pieces.slice(0, -1).join('\n')}\n${pieces.slice(-1)}`;
    },
    removeEmptyLines: (text: string): string => {
        return text.replace(/^\s*$(?:\r\n?|\n)/gm, "");
    },
    removeHorizontalWhiteSpace: (text: string): string => {
        const pieces: string[] = text.split("\n");
        let result: string[] = [];

        pieces.forEach((piece: string, i: number) => {
            piece = piece.replace(/^\s+|\s+$/g, '');
            result.push(piece);
        });

        return `${result.slice(0, -1).join('\n')}\n${result.slice(-1)}`;
    },
    removeDuplicateLines: (text: string): string => {
        const pieces: string[] = text.split("\n");
        let dummyArray: string[] = [];
        let index: number[] = [];
        let result: string[] = [];

        pieces.forEach((piece: string, i: number) => {
            piece = piece.replace(/^\s+|\s+$/g, '');
            piece = piece.toLowerCase();

            if (piece === '<br>' || piece === '<br/>' || piece === '<br />' || dummyArray.indexOf(piece) < 0) {
                index.push(i);
                dummyArray.push(piece);
            }
        });

        index.forEach((i: number) => { result.push(pieces[i]); });
        return `${result.slice(0, -1).join('\n')}\n${result.slice(-1)}`;
    }
};

src/helpers/commandHelpers.ts
import { IExtensionCommand } from "../interfaces/IExtensionCommand";
import { Utils } from "../Utils";

export class CommandHelpers {
    getAllCommands(): IExtensionCommand[] {
        let commands: IExtensionCommand[] = [];

        commands.push({ label: "HTML Encode", commandName: "htmlEncode" });
        commands.push({ label: "HTML Decode", commandName: "htmlDecode" });

        commands.push({ label: "URL Encode", commandName: "urlEncode" });
        commands.push({ label: "URL Decode", commandName: "urlDecode" });

        commands.push({ label: "To Title Case", commandName: "toTitleCase" });
        commands.push({ label: "To Lower Case", commandName: "toLowerCase" });
        commands.push({ label: "To Upper Case", commandName: "toUpperCase" });
        commands.push({ label: "Capitalize First Letter", commandName: "capitalizeFirstLetter" });
        commands.push({ label: "Reverse", commandName: "reverse" });
        commands.push({ label: "Remove Accents/Diacritics", commandName: "removeAccent" });
        commands.push({ label: "MD5", commandName: "toMd5" });
        commands.push({ label: "SHA-1", commandName: "toSha1" });
        commands.push({ label: "SHA-256", commandName: "toSha256" });
        commands.push({ label: "SHA-384", commandName: "toSha384" });
        commands.push({ label: "SHA-512", commandName: "toSha512" });

        commands.push({ label: "Sort Ascending", commandName: "sortAscending" });
        commands.push({ label: "Sort Descending", commandName: "sortDescending" });
        commands.push({ label: "Remove Empty Lines", commandName: "removeEmptyLines" });
        commands.push({ label: "Remove Horizontal White Space", commandName: "removeHorizontalWhiteSpace" });
        commands.push({ label: "Remove Duplicate Lines", commandName: "removeDuplicateLines" });

        return commands;
    }

    getCommandResult(selection: string, cmd: string): string {
        return Utils[cmd](selection);
    }
}